### DISCUSSION TITLE
Compliant with Microsoft SQL

### DISCUSSION THREAD

**Discussion by pepijn-hockers** [sqlc-dev/sqlc#1049] [Q&A]
Hi there,

I was investigating this package for possible usage within my service. However, I couldn't find a reference on Microsoft SQL server. Is the package compliant and usable for Microsoft SQL? MSSQL?

Kind regards,

Pepijn

**Comment by kyleconroy** [comment]
MSSQL support is planned but not implemented yet. Hopefully soon!

**Reply by tooolbox** [reply]
Out of curiosity, what would be necessary to achieve this?

1. Copy one of the existing folders in the engine directory, fiddle with it.
2. Add whatever command line flags or config to enable use of said directory under the right circumstances.
3. Profit?

I’m an experienced Go dev so the simplification is tongue-in-cheek, but I’m curious what the high-level is on this task.

**Reply by kyleconroy** [reply]
Adding a new database engine requires finding a high-quality parser for its SQL dialect. I don't know of any good MSSQL parsers written in Go.

**Reply by tooolbox** [reply]
Yes, I did a search and can’t find anything.  I guess the best option may be modifying an existing one such as the Vitess parser.

**Reply by tooolbox** [reply]
What am I saying?  Probably better to find an existing library in another language and port it.

**Reply by tooolbox** [reply]
Since I saw you are using an ANTLR v4 based parser for SQLite, I found such a grammar for tsql.  Not complete, but might be suitable as a starting point.

https://github.com/antlr/grammars-v4/tree/master/sql/tsql

**Reply by tooolbox** [reply]
There is also a pre-generated Go package for this grammar:

https://github.com/bramp/antlr4-grammars/tree/master/tsql

**Reply by tooolbox** [reply]
@kyleconroy in some spare time last year, I started building a native-go [TSQL parser](https://github.com/tooolbox/participle/tree/tsql/_examples/tsql) using [Participle](https://github.com/alecthomas/participle).  I got it working with [a large set of SELECT queries](https://github.com/tooolbox/participle/blob/tsql/_examples/tsql/testdata/dml_select.sql).

Obviously next on the agenda would be INSERT, UPDATE, DELETE.  I'm wondering what other statements would need to be parseable for the package to be useful to sqlc.

**Comment by fautore** [comment]
hi, what's the status on the Microsoft SQL compliance/support?

**Reply by kyleconroy** [reply]
Still waiting for a MSSQL parser written in Go

**Reply by tooolbox** [reply]
I was working on this a while back, got [this far](https://github.com/morphar/sqlparsers/pull/4) but no farther.

**Reply by koalazub** [reply]
@tooolbox last merge was 7 years ago but it seems like the conversation was still going with the author. Did the conversation continue elsewhere?

**Reply by tooolbox** [reply]
No, it did not continue elsewhere.  I suppose we were both busy and nothing further happened.

As of now I still don't have time to pursue that project further, but there is a parser for SELECT/UPDATE/INSERT/DELETE which passes a large number of test cases.  Anyone is welcome to pick it up and take if further; if I need to shove an MIT license in there I will.

**Comment by EliSauder** [comment]
For the parser what is the hesitancy to use https://github.com/antlr/grammars-v4/tree/master/sql/tsql as mentioned by @tooolbox?

@kyleconroy

**Reply by tooolbox** [reply]
I personally did not have a lot of luck trying to use that generated antlr parser when I tried, years ago.  Which is why I tried to write my own.

Looking at it now, there does appear to be the recent addition of: https://github.com/ha1tch/tsqlparser

**Reply by EliSauder** [reply]
Ah gotcha

**Reply by EliSauder** [reply]
@tooolbox are you familiar at all with https://github.com/microsoft/SqlScriptDOM and it's grammar files e.g. https://github.com/microsoft/SqlScriptDOM/blob/main/SqlScriptDom/Parser/TSql/TSql170.g

**Reply by tooolbox** [reply]
No, apologies, I am not.

As I said though, the one by ha1tch seems decent, no?  All sqlc needs is the AST, from my understanding.

**Reply by EliSauder** [reply]
Yeah, it looks pretty decent

**Comment by kyleconroy** [comment]
ANTLR grammars produce really gross ASTs. Our SQLite parser currently uses it and I want to avoid it. I've started workon https://github.com/sqlc-dev/teesql which is a parser written in Go. Try it out!

**Reply by tooolbox** [reply]
Are there downsides to https://github.com/ha1tch/tsqlparser as mentioned previously, such that one would need to start from scratch?

**Reply by kyleconroy** [reply]
It's a newer project with little usage, so I'd rather control the parser if I need to make any changes for sqlc.

**Reply by ha1tch** [reply]
Hey guys, this thread only came to my attention just now, I'm the author of ha1tch/tsqlparser.  Let me know if I can be of help with anything!
If you would like to assess its suitability for real world database engines, here are two of them:
https://github.com/ha1tch/xolu
https://github.com/ha1tch/aulsql

Xolu is in active development right now, and the T-SQL-oriented API is capable of fronting a SQLite-based backend. Work is ongoing to extend it to support a Postgres backend.
AulSQL is completely different and a more experimental at this time, but it serves to show that the parser can be used to keep filesystem-based T-SQL stored procedures that can be executed by a non-Microsoft runtime.  The point of this is to decouple stored procedures in such a way that they can easily be versioned in a Git repository (also running against a pluggable non-Microsoft backend).

Also relevant:
https://github.com/ha1tch/tgpiler
This is a transpiler from T-SQL to Go, using the same parser. This has been used in production in real world scenarios assisting a development team during a large scale migration project.

**Comment by 0xBradock** [comment]
Hello, is there any progress concerning MSSQL. I am happy to try and help with the implementation
